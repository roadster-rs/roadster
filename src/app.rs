use async_trait::async_trait;

use crate::config::app_config::AppConfig;
use crate::context::{AppContext, FromAppContextRef, GetAppContextFromRef};
use crate::tracing::init_tracing;

pub async fn start<A>() -> anyhow::Result<()>
where
    A: App + Default + Sync,
{
    let config = AppConfig::new()?;

    let app = A::default();
    app.init_tracing(&config)?;

    let context = AppContext::new(config).await?;
    let _context = app.context_to_state(&context).await?;

    Ok(())
}

#[async_trait]
pub trait App {
    type State: FromAppContextRef + GetAppContextFromRef;

    fn init_tracing(&self, config: &AppConfig) -> anyhow::Result<()> {
        init_tracing(config)?;

        Ok(())
    }

    async fn context_to_state(&self, app_context: &AppContext) -> anyhow::Result<Self::State> {
        let state = Self::State::from_app_context_ref(app_context).await?;
        Ok(*state)
    }
}
