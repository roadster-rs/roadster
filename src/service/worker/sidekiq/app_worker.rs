use crate::app::context::AppContext;
use crate::app::App;
use crate::error::RoadsterResult;
use async_trait::async_trait;
use serde_derive::{Deserialize, Serialize};
use serde_with::{serde_as, skip_serializing_none};
use sidekiq::Worker;
use std::time::Duration;
use typed_builder::TypedBuilder;
use validator::Validate;

/// Additional configuration options that can be configured via the app's configuration files.
/// The options can also be overridden on a per-worker basis by implementing the corresponding
/// method in the [AppWorker] trait.
#[serde_as]
#[skip_serializing_none]
#[derive(Debug, Clone, Validate, Serialize, Deserialize, TypedBuilder)]
#[serde(default, rename_all = "kebab-case")]
pub struct AppWorkerConfig {
    /// The maximum number of times a job should be retried on failure.
    #[builder(default = AppWorkerConfig::default().max_retries)]
    pub max_retries: usize,
    /// True if Roadster should enforce a timeout on the app's workers. The default duration of
    /// the timeout can be configured with the `max-duration` option.
    #[builder(default = AppWorkerConfig::default().timeout)]
    pub timeout: bool,
    /// The maximum duration workers should run for. The timeout is only enforced if `timeout`
    /// is `true`.
    #[serde_as(as = "serde_with::DurationSeconds")]
    #[builder(default = AppWorkerConfig::default().max_duration)]
    pub max_duration: Duration,
    /// See <https://docs.rs/rusty-sidekiq/latest/sidekiq/trait.Worker.html#method.disable_argument_coercion>
    #[builder(default = AppWorkerConfig::default().disable_argument_coercion)]
    pub disable_argument_coercion: bool,
}

impl Default for AppWorkerConfig {
    fn default() -> Self {
        AppWorkerConfig::builder()
            .max_retries(5)
            .timeout(true)
            .max_duration(Duration::from_secs(60))
            .disable_argument_coercion(false)
            .build()
    }
}

#[async_trait]
pub trait AppWorker<A, Args>: Worker<Args>
where
    Self: Sized,
    A: App,
    Args: Send + Sync + serde::Serialize + 'static,
{
    /// Build a new instance of the [worker][Self].
    fn build(context: &AppContext<A::State>) -> Self;

    /// Enqueue the worker into its Sidekiq queue. This is a helper method around [Worker::perform_async]
    /// so the caller can simply provide the [state][App::State] instead of needing to access the
    /// [sidekiq::RedisPool] from inside the [state][App::State].
    async fn enqueue(context: &AppContext<A::State>, args: Args) -> RoadsterResult<()> {
        Self::perform_async(context.redis_enqueue(), args).await?;
        Ok(())
    }

    /// Provide the [AppWorkerConfig] for [Self]. The default implementation populates the
    /// [AppWorkerConfig] using the values from the corresponding methods on [Self], e.g.,
    /// [Self::max_retries].
    fn config(&self, context: &AppContext<A::State>) -> AppWorkerConfig {
        AppWorkerConfig::builder()
            .max_retries(AppWorker::max_retries(self, context))
            .timeout(self.timeout(context))
            .max_duration(self.max_duration(context))
            .disable_argument_coercion(AppWorker::disable_argument_coercion(self, context))
            .build()
    }

    /// See [AppWorkerConfig::max_retries].
    ///
    /// The default implementation uses the value from the app's config file.
    fn max_retries(&self, context: &AppContext<A::State>) -> usize {
        context
            .config()
            .service
            .sidekiq
            .custom
            .app_worker
            .max_retries
    }

    /// See [AppWorkerConfig::timeout].
    ///
    /// The default implementation uses the value from the app's config file.
    fn timeout(&self, context: &AppContext<A::State>) -> bool {
        context.config().service.sidekiq.custom.app_worker.timeout
    }

    /// See [AppWorkerConfig::max_duration].
    ///
    /// The default implementation uses the value from the app's config file.
    fn max_duration(&self, context: &AppContext<A::State>) -> Duration {
        context
            .config()
            .service
            .sidekiq
            .custom
            .app_worker
            .max_duration
    }

    /// See [AppWorkerConfig::disable_argument_coercion].
    ///
    /// The default implementation uses the value from the app's config file.
    fn disable_argument_coercion(&self, context: &AppContext<A::State>) -> bool {
        context
            .config()
            .service
            .sidekiq
            .custom
            .app_worker
            .disable_argument_coercion
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::util::serde_util::Wrapper;
    use serde_json::from_str;

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_config_override_max_retries() {
        let max_retries = 1234;
        let value: Wrapper<AppWorkerConfig> = from_str(&format!(
            r#"{{"inner": {{"max-retries": {max_retries} }} }}"#
        ))
        .unwrap();
        assert_eq!(value.inner.max_retries, max_retries);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_config_override_timeout() {
        let value: Wrapper<AppWorkerConfig> =
            from_str(r#"{"inner": {"timeout": false } }"#).unwrap();
        assert!(!value.inner.timeout);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_config_override_max_duration() {
        let max_duration = Duration::from_secs(1234);
        let value: Wrapper<AppWorkerConfig> = from_str(&format!(
            r#"{{"inner": {{"max-duration": {} }} }}"#,
            max_duration.as_secs()
        ))
        .unwrap();
        assert_eq!(value.inner.max_duration, max_duration);
    }

    #[test]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn deserialize_config_override_disable_argument_coercion() {
        let value: Wrapper<AppWorkerConfig> =
            from_str(r#"{"inner": {"disable-argument-coercion": true } }"#).unwrap();
        assert!(value.inner.disable_argument_coercion);
    }
}

#[cfg(test)]
mod deserialize_tests {
    use super::*;
    use crate::util::test_util::TestCase;
    use insta::assert_toml_snapshot;
    use rstest::{fixture, rstest};

    #[fixture]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn case() -> TestCase {
        Default::default()
    }

    #[rstest]
    #[case("")]
    #[case(
        r#"
        max-retries = 1
        "#
    )]
    #[case(
        r#"
        timeout = false
        "#
    )]
    #[case(
        r#"
        max-duration = 1234
        "#
    )]
    #[case(
        r#"
        disable-argument-coercion = true
        "#
    )]
    #[cfg_attr(coverage_nightly, coverage(off))]
    fn app_worker(_case: TestCase, #[case] config: &str) {
        let app_worker: AppWorkerConfig = toml::from_str(config).unwrap();

        assert_toml_snapshot!(app_worker);
    }
}
