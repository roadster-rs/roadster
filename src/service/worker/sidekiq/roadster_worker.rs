use crate::app::context::AppContext;
use crate::service::worker::sidekiq::app_worker::AppWorkerConfig;
use crate::service::worker::sidekiq::app_worker::DEFAULT_MAX_DURATION;
use async_trait::async_trait;
use serde::Serialize;
use sidekiq::{RedisPool, Worker, WorkerOpts};
use std::marker::PhantomData;
use std::time::Duration;
use tracing::{error, instrument};

/// Worker used by Roadster to wrap the consuming app's workers to add additional behavior. For
/// example, [RoadsterWorker] is by default configured to automatically abort the app's worker
/// when it exceeds a certain timeout.
pub(crate) struct RoadsterWorker<Args, W>
where
    Args: Send + Sync + Serialize + 'static,
    W: Worker<Args>,
{
    inner: W,
    inner_config: Option<AppWorkerConfig>,
    context: AppContext,
    _args: PhantomData<Args>,
}

impl<Args, W> RoadsterWorker<Args, W>
where
    Args: Send + Sync + Serialize,
    W: Worker<Args>,
{
    pub(crate) fn new(context: &AppContext, inner: W, config: Option<AppWorkerConfig>) -> Self {
        Self {
            inner,
            inner_config: config,
            context: context.clone(),
            _args: PhantomData,
        }
    }
}

#[async_trait]
impl<Args, W> Worker<Args> for RoadsterWorker<Args, W>
where
    Args: Send + Sync + Serialize,
    W: Worker<Args>,
{
    fn disable_argument_coercion(&self) -> bool {
        self.inner_config
            .as_ref()
            .and_then(|config| config.disable_argument_coercion)
            .unwrap_or_else(|| {
                self.context
                    .config()
                    .service
                    .sidekiq
                    .custom
                    .app_worker
                    .disable_argument_coercion
                    .unwrap_or_else(|| W::disable_argument_coercion(&self.inner))
            })
    }

    fn opts() -> WorkerOpts<Args, Self>
    where
        Self: Sized,
    {
        // This method not implemented because `RoadsterWorker` should not be enqueued directly,
        // and this method is only used when enqueuing. Instead, Sidekiq.rs will use the
        // `W::opts` implementation directly.
        unimplemented!()
    }

    fn max_retries(&self) -> usize {
        self.inner_config
            .as_ref()
            .and_then(|config| config.max_retries)
            .unwrap_or_else(|| {
                self.context
                    .config()
                    .service
                    .sidekiq
                    .custom
                    .app_worker
                    .max_retries
                    .unwrap_or_else(|| W::max_retries(&self.inner))
            })
    }

    fn class_name() -> String
    where
        Self: Sized,
    {
        // This method is implemented because it's used both when registering the worker, and
        // when enqueuing a job. We forward the implementation to `W::classname` because that's
        // what Sidekiq.rs uses specifically. If we attempt to override this, our impl will be used
        // when registering the worker, but not when enqueuing a job, so the worker will not pick
        // up the jobs.
        W::class_name()
    }

    async fn perform_async(_redis: &RedisPool, _args: Args) -> sidekiq::Result<()>
    where
        Self: Sized,
        Args: Send + Sync + Serialize + 'static,
    {
        // This method not implemented because `RoadsterWorker` should not be enqueued directly.
        unimplemented!()
    }

    async fn perform_in(_redis: &RedisPool, _duration: Duration, _args: Args) -> sidekiq::Result<()>
    where
        Self: Sized,
        Args: Send + Sync + Serialize + 'static,
    {
        // This method not implemented because `RoadsterWorker` should not be enqueued directly.
        unimplemented!()
    }

    #[instrument(skip_all)]
    async fn perform(&self, args: Args) -> sidekiq::Result<()> {
        let inner = self.inner.perform(args);

        let timeout = self
            .inner_config
            .as_ref()
            .and_then(|config| config.timeout)
            .unwrap_or_else(|| {
                self.context
                    .config()
                    .service
                    .sidekiq
                    .custom
                    .app_worker
                    .timeout
                    .unwrap_or_default()
            });

        if timeout {
            let max_duration = self
                .inner_config
                .as_ref()
                .and_then(|config| config.max_duration)
                .unwrap_or_else(|| {
                    self.context
                        .config()
                        .service
                        .sidekiq
                        .custom
                        .app_worker
                        .max_duration
                        .unwrap_or(DEFAULT_MAX_DURATION)
                });

            tokio::time::timeout(max_duration, inner)
                .await
                .map_err(|err| {
                    error!(
                        worker = %W::class_name(),
                        max_duration = %max_duration.as_secs(),
                        %err,
                        "Worker timed out"
                    );
                    sidekiq::Error::Any(Box::new(err))
                })?
        } else {
            inner.await
        }
    }
}
