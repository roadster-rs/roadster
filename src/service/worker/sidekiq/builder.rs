use crate::app::App;
use crate::app_context::AppContext;
use crate::config::service::worker::sidekiq::StaleCleanUpBehavior;
use crate::service::worker::sidekiq::app_worker::AppWorker;
use crate::service::worker::sidekiq::roadster_worker::RoadsterWorker;
use crate::service::worker::sidekiq::service::SidekiqWorkerService;
use crate::service::{AppService, AppServiceBuilder};
use anyhow::{anyhow, bail};
use async_trait::async_trait;
use itertools::Itertools;
use num_traits::ToPrimitive;
use serde::Serialize;
use sidekiq::{periodic, Processor, ProcessorConfig};
use std::collections::HashSet;
use std::sync::Arc;
use tracing::{debug, info, warn};

const PERIODIC_KEY: &str = "periodic";

pub struct SidekiqWorkerServiceBuilder<A>
where
    A: App,
{
    state: BuilderState<A>,
}

enum BuilderState<A: App> {
    Enabled {
        processor: Processor,
        context: Arc<AppContext>,
        state: Arc<A::State>,
        registered_workers: HashSet<String>,
        registered_periodic_workers: HashSet<String>,
    },
    Disabled,
}

#[async_trait]
impl<A> AppServiceBuilder<A, SidekiqWorkerService> for SidekiqWorkerServiceBuilder<A>
where
    A: App + 'static,
{
    fn enabled(&self, app_context: &AppContext, app_state: &A::State) -> bool {
        match self.state {
            BuilderState::Enabled { .. } => {
                <SidekiqWorkerService as AppService<A>>::enabled(app_context, app_state)
            }
            BuilderState::Disabled => false,
        }
    }

    async fn build(
        self,
        context: &AppContext,
        _state: &A::State,
    ) -> anyhow::Result<SidekiqWorkerService> {
        let service = match self.state {
            BuilderState::Enabled {
                processor,
                registered_periodic_workers,
                ..
            } => {
                Self::remove_stale_periodic_jobs(context, &registered_periodic_workers).await?;
                SidekiqWorkerService { processor }
            }
            BuilderState::Disabled => {
                bail!("This builder is not enabled; it's build method should not have been called.")
            }
        };

        Ok(service)
    }
}

impl<A> SidekiqWorkerServiceBuilder<A>
where
    A: App + 'static,
{
    pub async fn with_processor(
        context: Arc<AppContext>,
        state: Arc<A::State>,
        processor: Processor,
    ) -> anyhow::Result<Self> {
        Self::new(context, state, Some(processor)).await
    }

    pub async fn with_default_processor(
        context: Arc<AppContext>,
        state: Arc<A::State>,
        worker_queues: Option<Vec<String>>,
    ) -> anyhow::Result<Self> {
        let processor = if !<SidekiqWorkerService as AppService<A>>::enabled(&context, &state) {
            debug!("Sidekiq service not enabled, not creating the Sidekiq processor");
            None
        } else if let Some(redis_fetch) = context.redis_fetch() {
            Self::auto_clean_periodic(&context).await?;
            let queues = context
                .config()
                .service
                .sidekiq
                .custom
                .queues
                .clone()
                .into_iter()
                .chain(worker_queues.unwrap_or_default())
                .collect_vec();
            info!(
                "Creating Sidekiq.rs (rusty-sidekiq) processor with {} queues",
                queues.len()
            );
            debug!("Sidekiq.rs queues: {queues:?}");
            let processor = {
                let num_workers = context
                    .config()
                    .service
                    .sidekiq
                    .custom
                    .num_workers
                    .to_usize()
                    .ok_or_else(|| {
                        anyhow!(
                            "Unable to convert num_workers `{}` to usize",
                            context.config().service.sidekiq.custom.num_workers
                        )
                    })?;
                let processor_config: ProcessorConfig = Default::default();
                let processor_config = processor_config.num_workers(num_workers);
                Processor::new(redis_fetch.clone(), queues.clone()).with_config(processor_config)
            };

            Some(processor)
        } else {
            debug!(
                "No 'redis fetch' connection pool configured, not creating the Sidekiq processor"
            );
            None
        };

        Self::new(context, state, processor).await
    }

    async fn new(
        context: Arc<AppContext>,
        state: Arc<A::State>,
        processor: Option<Processor>,
    ) -> anyhow::Result<Self> {
        let processor = if <SidekiqWorkerService as AppService<A>>::enabled(&context, &state) {
            processor
        } else {
            None
        };

        let state = if let Some(processor) = processor {
            BuilderState::Enabled {
                processor,
                context,
                state,
                registered_workers: Default::default(),
                registered_periodic_workers: Default::default(),
            }
        } else {
            BuilderState::Disabled
        };

        Ok(Self { state })
    }

    async fn auto_clean_periodic(context: &AppContext) -> anyhow::Result<()> {
        if context
            .config()
            .service
            .sidekiq
            .custom
            .periodic
            .stale_cleanup
            == StaleCleanUpBehavior::AutoCleanAll
        {
            // Periodic jobs are not removed automatically. Remove any periodic jobs that were
            // previously added. They should be re-added by `App::worker`.
            info!("Auto-cleaning periodic jobs");
            periodic::destroy_all(context.redis_enqueue().clone()).await?;
        }

        Ok(())
    }

    /// Remove previously-registered periodic jobs from Sidekiq/Redis. This should be called
    /// before registering any new periodic jobs. If this method is called after a periodic job is
    /// registered, it will return an error.
    ///
    /// Periodic jobs can also be cleaned up automatically by setting the
    /// [service.sidekiq.periodic.stale-cleanup][crate::config::service::worker::sidekiq::StaleCleanUpBehavior]
    /// to `auto-clean-all` or `auto-clean-stale`.
    pub async fn clean_up_periodic_jobs(self) -> anyhow::Result<Self> {
        if let BuilderState::Enabled {
            registered_periodic_workers,
            context,
            ..
        } = &self.state
        {
            if !registered_periodic_workers.is_empty() {
                bail!("Can only clean up previous periodic jobs if no periodic jobs have been registered yet.")
            }
            periodic::destroy_all(context.redis_enqueue().clone()).await?;
        }

        Ok(self)
    }

    /// Register a [worker][AppWorker] to handle Sidekiq.rs jobs.
    ///
    /// The worker will be wrapped by a [RoadsterWorker], which provides some common behavior, such
    /// as enforcing a timeout/max duration of worker jobs.
    pub fn register_app_worker<Args, W>(mut self, worker: W) -> anyhow::Result<Self>
    where
        Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
        W: AppWorker<A, Args> + 'static,
    {
        if let BuilderState::Enabled {
            processor,
            state,
            registered_workers,
            ..
        } = &mut self.state
        {
            let class_name = W::class_name();
            debug!(worker = %class_name, "Registering worker");
            if !registered_workers.insert(class_name.clone()) {
                bail!("Worker `{class_name}` was already registered");
            }
            let roadster_worker = RoadsterWorker::new(worker, state.clone());
            processor.register(roadster_worker);
        }

        Ok(self)
    }

    /// Register a periodic [worker][AppWorker] that will run with the provided args. The cadence
    /// of the periodic worker, the worker's queue name, and other attributes are specified using
    /// the [builder][periodic::Builder]. However, to help ensure type-safety the args are provided
    /// to this method instead of the [builder][periodic::Builder].
    ///
    /// The worker will be wrapped by a [RoadsterWorker], which provides some common behavior, such
    /// as enforcing a timeout/max duration of worker jobs.
    pub async fn register_periodic_app_worker<Args, W>(
        mut self,
        builder: periodic::Builder,
        worker: W,
        args: Args,
    ) -> anyhow::Result<Self>
    where
        Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
        W: AppWorker<A, Args> + 'static,
    {
        if let BuilderState::Enabled {
            processor,
            state,
            registered_periodic_workers,
            ..
        } = &mut self.state
        {
            let class_name = W::class_name();
            debug!(worker = %class_name, "Registering periodic worker");
            let roadster_worker = RoadsterWorker::new(worker, state.clone());
            let builder = builder.args(args)?;
            let job_json = serde_json::to_string(&builder.into_periodic_job(class_name.clone())?)?;
            if !registered_periodic_workers.insert(job_json.clone()) {
                bail!(
                    "Periodic worker `{class_name}` was already registered; full job: {job_json}"
                );
            }
            builder.register(processor, roadster_worker).await?;
        }

        Ok(self)
    }

    /// Compares the list of periodic jobs that were registered by the app during app startup with
    /// the list of periodic jobs in Redis, and removes any that exist in Redis but weren't
    /// registered during start up.
    ///
    /// The jobs are only removed if the [worker.sidekiq.periodic.stale-cleanup][crate::config::worker::Periodic]
    /// config is set to [auto-clean-stale][StaleCleanUpBehavior::AutoCleanStale].
    ///
    /// This is run after all the app's periodic jobs have been registered.
    pub(crate) async fn remove_stale_periodic_jobs(
        context: &AppContext,
        registered_periodic_workers: &HashSet<String>,
    ) -> anyhow::Result<()> {
        let mut conn = context.redis_enqueue().get().await?;
        let stale_jobs = conn
            .zrange(PERIODIC_KEY.to_string(), 0, -1)
            .await?
            .into_iter()
            .filter(|job| !registered_periodic_workers.contains(job))
            .collect_vec();

        if stale_jobs.is_empty() {
            info!("No stale periodic jobs found");
            return Ok(());
        }

        if context
            .config()
            .service
            .sidekiq
            .custom
            .periodic
            .stale_cleanup
            == StaleCleanUpBehavior::AutoCleanStale
        {
            info!(
                "Removing {} stale periodic jobs:\n{}",
                stale_jobs.len(),
                stale_jobs.join("\n")
            );
            conn.zrem(PERIODIC_KEY.to_string(), &stale_jobs).await?;
        } else {
            warn!(
                "Found {} stale periodic jobs:\n{}",
                stale_jobs.len(),
                stale_jobs.join("\n")
            );
        }

        Ok(())
    }
}
