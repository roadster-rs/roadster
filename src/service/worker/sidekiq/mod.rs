use crate::app::App;
use crate::service::worker::sidekiq::app_worker::AppWorker;
use crate::service::worker::sidekiq::roadster_worker::RoadsterWorker;
use serde::Serialize;
use sidekiq::periodic;
use std::marker::PhantomData;

pub mod app_worker;
pub mod builder;
pub mod roadster_worker;
pub mod service;

/// A wrapper around [sidekiq::Processor] to help with mocking; we can't simply mock
/// sidekiq::Processor because [periodic::Builder] takes a [sidekiq::Processor] in order
/// to register a periodic job, so it won't be albe to take a MockProcessor created by `mockall`.
#[derive(Clone)]
struct Processor<A>
where
    A: App + 'static,
{
    inner: sidekiq::Processor,
    _app: PhantomData<A>,
}

impl<A> Processor<A>
where
    A: App + 'static,
{
    #[cfg_attr(test, allow(dead_code))]
    fn new(inner: sidekiq::Processor) -> Self {
        Self {
            inner,
            _app: PhantomData,
        }
    }

    #[cfg_attr(test, allow(dead_code))]
    fn register<Args, W>(&mut self, worker: RoadsterWorker<A, Args, W>)
    where
        Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
        W: AppWorker<A, Args> + 'static,
    {
        self.inner.register(worker);
    }

    #[cfg_attr(test, allow(dead_code))]
    async fn register_periodic<Args, W>(
        &mut self,
        builder: periodic::Builder,
        worker: RoadsterWorker<A, Args, W>,
    ) -> anyhow::Result<()>
    where
        Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
        W: AppWorker<A, Args> + 'static,
    {
        builder.register(&mut self.inner, worker).await?;
        Ok(())
    }

    #[cfg_attr(test, allow(dead_code))]
    fn into_sidekiq_processor(self) -> sidekiq::Processor {
        self.inner
    }
}

#[cfg(test)]
mockall::mock! {
    Processor<A: App + 'static> {
        fn new(inner: sidekiq::Processor) -> Self;

        fn register<Args, W>(&mut self, worker: RoadsterWorker<A, Args, W>)
        where
            Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
            W: AppWorker<A, Args> + 'static;

        async fn register_periodic<Args, W>(
            &mut self,
            builder: periodic::Builder,
            worker: RoadsterWorker<A, Args, W>,
        ) -> anyhow::Result<()>
        where
            Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
            W: AppWorker<A, Args> + 'static;

        fn into_sidekiq_processor(self) -> sidekiq::Processor;
    }

    impl<A: App + 'static> Clone for Processor<A> {
        fn clone(&self) -> Self;
    }
}
