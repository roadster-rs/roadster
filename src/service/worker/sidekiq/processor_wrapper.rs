use crate::error::RoadsterResult;
use crate::service::worker::sidekiq::roadster_worker::RoadsterWorker;
use serde::Serialize;
use sidekiq::{periodic, ServerMiddleware, Worker};

/// A wrapper around [sidekiq::Processor] to help with mocking; we can't simply mock
/// sidekiq::Processor because [periodic::Builder] takes a [sidekiq::Processor] in order
/// to register a periodic job, so it won't be albe to take a MockProcessor created by `mockall`.
#[derive(Clone)]
pub(crate) struct ProcessorWrapper {
    inner: sidekiq::Processor,
}

impl ProcessorWrapper {
    #[cfg_attr(test, allow(dead_code))]
    pub(crate) fn new(inner: sidekiq::Processor) -> Self {
        Self { inner }
    }

    #[cfg_attr(test, allow(dead_code))]
    pub(crate) fn register<Args, W>(&mut self, worker: RoadsterWorker<Args, W>)
    where
        Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
        W: Worker<Args> + 'static,
    {
        self.inner.register(worker);
    }

    #[cfg_attr(test, allow(dead_code))]
    pub(crate) async fn register_periodic<Args, W>(
        &mut self,
        builder: periodic::Builder,
        worker: RoadsterWorker<Args, W>,
    ) -> RoadsterResult<()>
    where
        Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
        W: Worker<Args> + 'static,
    {
        builder.register(&mut self.inner, worker).await?;
        Ok(())
    }

    #[cfg_attr(test, allow(dead_code))]
    pub(crate) async fn middleware<M>(&mut self, middleware: M)
    where
        M: ServerMiddleware + Send + Sync + 'static,
    {
        self.inner.using(middleware).await;
    }

    #[cfg_attr(test, allow(dead_code))]
    pub(crate) fn into_sidekiq_processor(self) -> sidekiq::Processor {
        self.inner
    }
}

#[cfg(test)]
mockall::mock! {
    pub(crate) ProcessorWrapper {
        pub(crate) fn new(inner: sidekiq::Processor) -> Self;

        pub(crate) fn register<Args, W>(&mut self, worker: RoadsterWorker<Args, W>)
        where
            Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
            W: Worker<Args> + 'static;

        pub(crate) async fn register_periodic<Args, W>(
            &mut self,
            builder: periodic::Builder,
            worker: RoadsterWorker<Args, W>,
        ) -> RoadsterResult<()>
        where
            Args: Sync + Send + Serialize + for<'de> serde::Deserialize<'de> + 'static,
            W: Worker<Args> + 'static;

        pub(crate) async fn middleware<M>(&mut self, middleware: M)
        where
            M: ServerMiddleware + Send + Sync + 'static;

        pub(crate) fn into_sidekiq_processor(self) -> sidekiq::Processor;
    }

    impl Clone for ProcessorWrapper {
        fn clone(&self) -> Self;
    }
}
