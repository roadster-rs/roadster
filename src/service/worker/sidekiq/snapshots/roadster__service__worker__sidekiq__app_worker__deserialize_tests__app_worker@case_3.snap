---
source: src/service/worker/sidekiq/app_worker.rs
expression: app_worker
---
max-retries = 5
timeout = false
max-duration = 60
disable-argument-coercion = false
